### PR TITLE
Fix data file publishing

### PR DESCRIPTION
**Pull request by Matyrobbrt**
Data files weren't being published as artifacts to maven

### CODE REVIEW

File: testproject/common/build.gradle [hunk 1]
```diff
@@ -19,3 +20,16 @@ neoForge {
         }
     }
 }
+
+publishing {
+    publications {
+        yes(MavenPublication) {
```
Comment (Technici4n) [inline testproject/common/build.gradle:26]: People might actually reference this, please no meme names.

File: src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java [hunk 2]
```diff
@@ -1001,9 +1001,15 @@ private static Configuration dataFileConfiguration(Project project, String name,
             spec.withDependencies(dependencies -> dependencies.add(depFactory.create(collection.getPublished())));
         });
 
-        // Set up the publishing conditionally
-        AdhocComponentWithVariants java = (AdhocComponentWithVariants) project.getComponents().getByName("java");
+        // Publish the data files as artifacts whose classifiers match the name of the file to avoid conflicts
+        // Unfortunately, ArtifactHandler does not have lazy methods, so afterEval it is
+        project.afterEvaluate(proj -> collection.getPublished()
+                .forEach(file -> proj.getArtifacts().add(elementsConfiguration.getName(), file, configurablePublishArtifact -> {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java:1007]: Remove forEach? (Lambda soup...)

**Review by Technici4n — CHANGES_REQUESTED** [review]
I don't think we are quite there. The `common` project is publishing an empty `interfaceInjectionDataElements` variant.

File: src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java [hunk 3]
```diff
@@ -1013,7 +1022,7 @@ private static Configuration dataFileConfiguration(Project project, String name,
         };
 
         elementsConfiguration.getAllArtifacts().configureEach(artifact -> configurePublishing.run());
-        elementsConfiguration.getArtifacts().configureEach(artifact -> configurePublishing.run());
+        elementsConfiguration.getAllDependencies().configureEach(dep -> configurePublishing.run());
```
Comment (Technici4n) [inline src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java:1025]: image

Looks like this is always called because of `dependencies.add(depFactory.create(collection.getPublished()))` above. That counts as a dependency, and there is no resolution that can find out that it's empty.

File: src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java [hunk 4]
```diff
@@ -1001,9 +1001,18 @@ private static Configuration dataFileConfiguration(Project project, String name,
...
         });
 
-        // Set up the publishing conditionally
-        AdhocComponentWithVariants java = (AdhocComponentWithVariants) project.getComponents().getByName("java");
+        // Publish the data files as artifacts whose classifiers match the name of the file to avoid conflicts
+        // Unfortunately, ArtifactHandler does not have lazy methods, so afterEval it is
+        project.afterEvaluate(proj -> {
+            for (var file : collection.getPublished()) {
+                proj.getArtifacts().add(elementsConfiguration.getName(), file, configurablePublishArtifact -> {
+                    configurablePublishArtifact.setClassifier(FileUtils.stripExtension(file.getName()));
```
Comment (Technici4n) [inline src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java:1009]: I suppose that Gradle already finds out about the extension? What is the default classifier that it chooses then? (i.e. Why do we need to override the classifier and strip the extension?).

File: testproject/subproject/build.gradle [hunk 5]
```diff
@@ -20,3 +21,16 @@ dependencies {
     implementation(accessTransformers(project(':common')))
     accessTransformersElements(project(':common'))
 }
+
+publishing {
```
Comment (Technici4n) [inline testproject/subproject/build.gradle:25]: This is good for manual testing, but I wonder if it would be possible and worthwhile to have an automated integration test as well?

File: src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java (hunk 4 above)
Comment (Matyrobbrt) [reply src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java:1009]: It doesn't chose anything, for the subproject with an iface injection data file for instance, the artifact is `subproject-<version>.json`.

**Comment by Matyrobbrt** [comment]
Yeah I noticed, it's because the elements configuration always has a dependency, even if it's empty. I don't think it really matters in the end if the empty variant is published, i might just remove the "conditional" part which is basically always true.

File: testproject/subproject/build.gradle (hunk 5 above)
Comment (Matyrobbrt) [reply testproject/subproject/build.gradle:25]: That sounds painful

**Comment by Technici4n** [comment]
I don't want to publish a variant for a niche feature by default, it's not great.

**Comment by Matyrobbrt** [comment]
And what do you suggest instead?

**Comment by Matyrobbrt** [comment]
Well, this is ugly, but I suppose it works:
```java
        elementsConfiguration.getAllDependencies().configureEach(dep -> {
            // We always add a dependency containing the project's published data files
            // Make sure that doesn't cause an empty variant
            if (dep instanceof FileCollectionDependency fc && fc.getFiles().isEmpty()) return;

            configurePublishing.run();
        });
```

File: src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java [hunk 6]
```diff
@@ -984,41 +995,58 @@ private static void addEclipseLaunchConfiguration(Project project,
...
-            spec.withDependencies(dependencies -> dependencies.add(depFactory.create(collection.getPublished())));
+            spec.attributes(attributes -> attributes.attribute(Category.CATEGORY_ATTRIBUTE, project.getObjects().named(Category.class, category)));
         });
 
-        // Set up the publishing conditionally
-        AdhocComponentWithVariants java = (AdhocComponentWithVariants) project.getComponents().getByName("java");
+        // Set up the variant publishing conditionally
+        var java = (AdhocComponentWithVariants) project.getComponents().getByName("java");
+        java.addVariantsFromConfiguration(elementsConfiguration, variant -> {
+            if (variant.getConfigurationVariant().getArtifacts().isEmpty()) {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java:1019]: Are we sure that this is lazy?

File: src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java [hunk 7]
```diff
@@ -984,41 +995,58 @@ private static void addEclipseLaunchConfiguration(Project project,
...
+            @Override
+            public void accept(Object artifactNotation) {
+                elementsConfiguration.getDependencies().add(depFactory.create(project.files(artifactNotation)));
+                project.getArtifacts().add(elementsConfiguration.getName(), artifactNotation, artifact -> {
+                    if (firstArtifact == null) {
+                        firstArtifact = artifact;
+                        artifact.setClassifier(category);
+                        artifactCount = 1;
+                    } else if (artifactCount == 1) {
+                        firstArtifact.setClassifier(category);
```
Comment (Technici4n) [inline src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java:1038]: Why change the classifier of only the first artifact? Shouldn't this change `artifact`?

File: src/test/java/net/neoforged/moddevgradle/functional/PublishDataFilesTest.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+                .withPluginClasspath()
+                .withProjectDir(testProjectDir)
+                .withArguments("printFileList")
+                .withDebug(true)
+                .build();
+
+        System.out.println(consumeResult.getOutput());
+
+    }
+
```
Comment (Technici4n) [inline src/test/java/net/neoforged/moddevgradle/functional/PublishDataFilesTest.java:158]: Thoughts on adding a test that makes sure we are not publishing empty variants?

File: src/main/java/net/neoforged/moddevgradle/dsl/DataFileCollection.java [hunk 9]
```diff
@@ -1,36 +1,49 @@
...
 
     /**
      * {@return the files this collection contains}
      */
     public abstract ConfigurableFileCollection getFiles();
-
-    /**
-     * {@return the files that should be published and that can be consumed by dependents}
-     */
-    public abstract ConfigurableFileCollection getPublished();
```
[the remark this answers is not in the document]
Comment (Technici4n) [reply src/main/java/net/neoforged/moddevgradle/dsl/DataFileCollection.java:35]: Nothing. Just return a dummy CFC and log a deprecation. :smile:

File: src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java (hunk 7 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java]: Ah yes, but `artifact` does not have its classifier set here. You are missing something like `artifact.setClassified(category + 2)`.

**Review by Technici4n — APPROVED** [review]
